arcgis-ios-sdk-samples/Route and directions/Display device location with NMEA data sources/SimulatedNMEADataSource.swift

```diff
@@ -0,0 +1,104 @@
+// You may obtain a copy of the License at
+//
+//   https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+protocol SimulatedNMEADataSourceDelegate: AnyObject {
+    /// A mock delegate method to receive NMEA data updates.
+    /// Replace it with the real one to receive NMEA sentences from a GPS dongle.
+    /// - Parameter nmeaData: The latest NMEA data.
+    func dataSource(didUpdate nmeaData: Data)
+}
+
+class SimulatedNMEADataSource {
+    /// The playback speed multiplier.
+    private let playbackSpeed: Double
+    /// An iterator to hold and loop through the mock NMEA data.
+    private var nmeaDataIterator: CircularIterator<Data>
+    /// A timer to periodically update NMEA data.
+    private var timer: Timer?
+    /// A delegate to notify view controllers.
+    weak var delegate: SimulatedNMEADataSourceDelegate?
+    
+    /// Load locations from NMEA sentences.
+    /// Read mock NMEA sentences line by line and group them by the timestamp.
+    /// - Parameters:
+    ///   - nmeaSourceFile: The URL of the NMEA source file.
+    ///   - speed: The playback speed multiplier.
+    init(from nmeaSourceFile: URL, speed: Double = 1.0) {
+        // An empty container for NMEA data.
+        var dataBySeconds = [Data]()
+        
+        if let nmeaStrings = try? String(contentsOf: nmeaSourceFile, encoding: .utf8).components(separatedBy: .newlines).filter({ !$0.isEmpty }) {
+            // A temporary container for the NMEA sentences at current timestamp.
+            var currentTimestamp = [String]()
+            for nmeaLine in nmeaStrings {
+                currentTimestamp.append(nmeaLine)
+                // In the mock data file, the sentences in each second end with
+                // an RMC message. Join all the messages in this second to a
+                // single data object.
+                let sentenceIdentifier = nmeaLine.prefix(6)
+                if sentenceIdentifier == "$GPRMC",
+                   !currentTimestamp.isEmpty {
+                    dataBySeconds.append(Data(currentTimestamp.joined(separator: "\n").utf8))
+                    currentTimestamp.removeAll()
+                }
+            }
+        }
+        // Create an iterator for the mock data generation.
+        nmeaDataIterator = CircularIterator(elements: dataBySeconds)
+        playbackSpeed = speed
+    }
+    
+    func start() {
+        guard !nmeaDataIterator.elements.isEmpty else { return }
+        // Invalidate timer to stop previous mock data generation.
+        timer?.invalidate()
+        // Duration in second.
+        let duration = TimeInterval(1) / playbackSpeed
```

Comment:
This actually creates an integer first, and then a time interval (double) from that integer:

![image](https://user-images.githubusercontent.com/2257493/109368504-306a7f00-784e-11eb-8f3c-0557994a541f.png)

```suggestion
        let duration: TimeInterval = 1 / playbackSpeed
```

However, I would also suggest renaming `duration`. The value is used as the time interval of the timer, so `interval` might be a better name.